packages/hebao_v3/contracts/base/libwallet/MetaTxLib.sol

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../lib/AddressUtil.sol";
+import "../../lib/EIP712.sol";
+import "../../lib/ERC20.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/SignatureUtil.sol";
+import "../../thirdparty/BytesUtil.sol";
+import "./WalletData.sol";
+import "./ERC20Lib.sol";
+import "./QuotaLib.sol";
+
+
+/// @title MetaTxLib
+/// @dev A module to support wallet meta-transactions.
+library MetaTxLib
+{
+    using AddressUtil   for address;
+    using BytesUtil     for bytes;
+    using MathUint      for uint;
+    using SignatureUtil for bytes32;
+    using QuotaLib      for Wallet;
+    using ERC20Lib      for Wallet;
+
+    bytes32 public constant META_TX_TYPEHASH = keccak256(
+        "MetaTx(address relayer,address to,uint256 nonce,address gasToken,uint256 gasPrice,uint256 gasLimit,uint256 gasOverhead,bytes data)"
+    );
+
+    event MetaTxExecuted(
+        address relayer,
+        uint    nonce,
+        bool    success,
+        uint    gasUsed
+    );
+
+    struct MetaTx
+    {
+        address to;
+        uint    nonce;
+        address gasToken;
+        uint    gasPrice;
+        uint    gasLimit;
+        uint    gasOverhead;
+        bool    requiresSuccess;
+        bytes   data;
+        bytes   signature;
+    }
+
+    function validateMetaTx(
+        Wallet  storage wallet,
+        bytes32 DOMAIN_SEPARATOR,
+        MetaTx  memory  metaTx
+        )
+        public
+        view
+    {
+        bytes memory encoded = abi.encode(
+            META_TX_TYPEHASH,
+            msg.sender,
+            metaTx.to,
+            metaTx.nonce,
+            metaTx.gasToken,
+            metaTx.gasPrice,
+            metaTx.gasLimit,
+            metaTx.gasOverhead,
+            metaTx.requiresSuccess,
+            keccak256(metaTx.data)
+        );
+        bytes32 metaTxHash = EIP712.hashPacked(DOMAIN_SEPARATOR, encoded);
+        require(metaTxHash.verifySignature(wallet.owner, metaTx.signature), "INVALID_SIGNATURE");
+    }
+
+    function executeMetaTx(
+        Wallet      storage wallet,
+        bytes32             DOMAIN_SEPARATOR,
+        PriceOracle         priceOracle,
+        MetaTx      memory  metaTx
+        )
+        public
+        returns (bool success)
+    {
+        require(!wallet.locked, "WALLET_LOCKED");
```

Comment:
If we require the wallet must be unlocked, then we cannot use meta-tx to unlock a wallet...
```

Reply:
Oh I guess it could be useful if the unlock is indeed done on the wallet itself, which also means the wallet owner can be trusted to sign the meta-tx itself (otherwise there would be no unlock). I think we can also allow that specific use-case by moving the `require(!wallet.locked, "WALLET_LOCKED");` check after the meta-tx call, perhaps at some extra risk to the relayer.

Reply:
Like this idea.